src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,411 @@
+                    if (_stateProvider.AccountExists(recipient))
+                    {
+                        if ((_virtualMachine.GetCachedCodeInfo(recipient)?.MachineCode?.Length ?? 0) != 0 || _stateProvider.GetNonce(recipient) != 0)
+                        {
+                            if (_logger.IsEnabled(LogEventLevel.Verbose))
+                            {
+                                _logger.Verbose($"Contract collision at {recipient}"); // the account already owns the contract with the code
+                            }
+
+                            throw new TransactionCollisionException();
+                        }
+
+                        _stateProvider.UpdateStorageRoot(recipient, Keccak.EmptyTreeHash);
+                    }
+                }
+
+                bool isPrecompile = recipient.IsPrecompiled(spec);
+
+                ExecutionEnvironment env = new ExecutionEnvironment();
+                env.Value = value;
+                env.TransferValue = value;
+                env.Sender = sender;
+                env.CodeSource = recipient;
+                env.ExecutingAccount = recipient;
+                env.CurrentBlock = stateUpdate;
+                env.GasPrice = gasPrice;
+                env.InputData = data ?? new byte[0];
+                env.CodeInfo = isPrecompile ? new CodeInfo(recipient) : machineCode == null ? _virtualMachine.GetCachedCodeInfo(recipient) : new CodeInfo(machineCode);
+                env.Originator = sender;
+
+                ExecutionType executionType = entry.IsValidDeploymentEntry ? ExecutionType.Create : ExecutionType.Call;
+                using (VmState state = new VmState(unspentGas, env, executionType, isPrecompile, true, false))
+                {
+                    substate = _virtualMachine.Run(state, txTracer);
+                    unspentGas = state.GasAvailable;
+                }
+
+                if (substate.ShouldRevert || substate.IsError)
+                {
+                    if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose("Restoring state from before transaction");
+                    _stateProvider.Restore(stateSnapshot);
+                    _storageProvider.Restore(storageSnapshot);
+                }
+                else
+                {
+                    // tks: there is similar code fo contract creation from init and from CREATE
+                    // this may lead to inconsistencies (however it is tested extensively in blockchain tests)
+                    if (entry.IsValidDeploymentEntry)
+                    {
+                        long codeDepositGasCost = CodeDepositHandler.CalculateCost(substate.Output.Length, spec);
+                        if (unspentGas < codeDepositGasCost && spec.IsEip2Enabled)
+                        {
+                            throw new OutOfGasException();
+                        }
+
+                        if (unspentGas >= codeDepositGasCost)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this code to not nest more than 3 control flow statements.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)